src/Symfony/Component/DependencyInjection/Compiler/MergeExtensionConfigurationPass.php

```diff
@@ -59,6 +60,18 @@ public function process(ContainerBuilder $container)
                 $tmpContainer->addObjectResource($configuration);
             }
 
+            if ($resolvingBag instanceof EnvPlaceholderParameterBag && method_exists(BaseNode::class, 'setPlaceholder')) {
+                $envTypes = $resolvingBag->getProvidedTypes();
+                foreach ($resolvingBag->getEnvPlaceholders() as $env => $placeholders) {
+                    if ((false === $i = strpos($env, ':')) || !isset($envTypes[$prefix = substr($env, 0, $i)])) {
+                        continue;
+                    }
+                    foreach ($placeholders as $placeholder) {
+                        BaseNode::setPlaceholder($placeholder, reset($envTypes[$prefix]));
```

Comment:
What about `getParam("env($env)")` as default value instead, if available :)

Reply:
Let's do. We also need to validate the type of the default value before calling setPlaceholder (its type needs to be in $envTypes).

Reply:
BUT, calling "reset" here prevents taking advantage of the "multiple types provided" info. So maybe pass an array instead, $type=>$value?